File: cpu.py
import json
import base64
from collections import deque
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError

TEAM_ID = "HGZZC"
AUTH_KEY = "ba0bf0fe-1197-48bd-943f-ce86dccd4435"
URL = "https://ctf26.vanierhacks.net/ALSS/cpu1"

def make_headers():
    token = base64.b64encode(f"{TEAM_ID}:{AUTH_KEY}".encode()).decode()
    return {
        "Authorization": f"Basic {token}",
        "Content-Type": "application/json",
        "Accept": "application/json",
    }

def fetch_payload():
    req = Request(URL, headers=make_headers(), method="GET")
    with urlopen(req) as resp:
        return json.load(resp)

def simulate_cpu1(challenge):
    regs = [0] * 8
    input_q = deque(challenge["input"])
    output_q = []

    for instr in challenge["instructions"]:
        opcode = (instr >> 8) & 0b11
        src = (instr >> 4) & 0xF
        dst = instr & 0xF

        if opcode != 0b01:
            raise ValueError(f"Unexpected opcode in cpu1: {opcode:02b} (instruction {instr})")

        # Read source
        if src <= 7:
            value = regs[src]
        elif src == 8:
            if not input_q:
                raise ValueError("Tried to read from input, but input queue is empty")
            value = input_q.popleft()
        else:
            raise ValueError(f"Unsupported source address for cpu1: {src}")

        value &= 0xFF

        # Write destination
        if dst <= 7:
            regs[dst] = value
        elif dst == 8:
            output_q.append(value)
        else:
            raise ValueError(f"Unsupported destination address for cpu1: {dst}")

    return output_q

def build_answer(payload):
    return {
        "submissionId": payload["submissionId"],
        "output": [simulate_cpu1(ch) for ch in payload["challenges"]]
    }

def post_answer(answer):
    data = json.dumps(answer).encode()
    req = Request(URL, data=data, headers=make_headers(), method="POST")
    with urlopen(req) as resp:
        return resp.read().decode()

def main():
    try:
        payload = fetch_payload()
        answer = build_answer(payload)

        print("Computed answer JSON:")
        print(json.dumps(answer, indent=2))

        print("\nPosting answer...\n")
        result = post_answer(answer)
        print("Server response:")
        print(result)

    except HTTPError as e:
        body = e.read().decode(errors="replace")
        print(f"HTTPError {e.code}")
        print(body)
    except URLError as e:
        print("URLError:", e)
    except Exception as e:
        print("Error:", e)

if __name__ == "__main__":
    main()

File: cpu.js
const http = require("http");
const fs = require("fs");

const TEAM_ID = "HGZZC";
const AUTH_KEY = "ba0bf0fe-1197-48bd-943f-ce86dccd4435";

function authHeader() {
  return "Basic " + Buffer.from(`${TEAM_ID}:${AUTH_KEY}`, "utf8").toString("base64");
}

function httpRequest(method, path, bodyObj = null) {
  return new Promise((resolve, reject) => {
    const body = bodyObj ? Buffer.from(JSON.stringify(bodyObj), "utf8") : null;

    const req = http.request(
      {
        hostname: "ctf26.vanierhacks.net",
        port: 80,
        path,
        method,
        headers: {
          Authorization: authHeader(),
          Accept: "application/json",
          ...(body
            ? {
                "Content-Type": "application/json; charset=utf-8",
                "Content-Length": body.length,
              }
            : {}),
          Connection: "close",
        },
      },
      (res) => {
        let data = "";
        res.setEncoding("utf8");

        res.on("data", (chunk) => {
          data += chunk;
        });

        res.on("end", () => {
          resolve({
            status: res.statusCode,
            headers: res.headers,
            body: data,
          });
        });
      }
    );

    req.on("error", reject);

    if (body) req.write(body);
    req.end();
  });
}

function simulateCPU1(challenge) {
  const regs = new Array(8).fill(0);
  const input = [...challenge.input];
  const output = [];

  for (const instr of challenge.instructions) {
    const opcode = (instr >> 8) & 0b11;
    const src = (instr >> 4) & 0xF;
    const dst = instr & 0xF;

    if (opcode !== 0b01) {
      throw new Error(`Unexpected opcode ${opcode} for instruction ${instr}`);
    }

    let value;

    if (src >= 0 && src <= 7) {
      value = regs[src];
    } else if (src === 8) {
      if (input.length === 0) {
        throw new Error("Tried to read from empty input queue");
      }
      value = input.shift();
    } else {
      throw new Error(`Unsupported source address ${src}`);
    }

    value &= 0xff;

    if (dst >= 0 && dst <= 7) {
      regs[dst] = value;
    } else if (dst === 8) {
      output.push(value);
    } else {
      throw new Error(`Unsupported destination address ${dst}`);
    }
  }

  return output;
}

async function main() {
  try {
    console.log("Fetching fresh challenge...");
    const getRes = await httpRequest("GET", "/ALSS/cpu1");

    console.log("GET status:", getRes.status);
    console.log("GET body:", getRes.body.slice(0, 500) + (getRes.body.length > 500 ? "..." : ""));

    if (getRes.status < 200 || getRes.status >= 300) {
      throw new Error(`GET failed with status ${getRes.status}`);
    }

    const payload = JSON.parse(getRes.body);

    const answer = {
      submissionId: String(payload.submissionId),
      output: payload.challenges.map((challenge) =>
        simulateCPU1(challenge).map((n) => Number(n))
      ),
    };

    fs.writeFileSync("answer.json", JSON.stringify(answer, null, 2));
    console.log("Wrote answer.json");
    console.log("submissionId:", answer.submissionId);
    console.log("challenge count:", answer.output.length);
    console.log("first output length:", answer.output[0]?.length ?? 0);

    console.log("\nSubmitting answer...");
    const postRes = await httpRequest("POST", "/ALSS/cpu1", answer);

    console.log("POST status:", postRes.status);
    console.log("POST body:", postRes.body);
  } catch (err) {
    console.error("Error:", err.message);
  }
}

main();